test/dataset/split/test_split.py

```diff
@@ -39,50 +39,145 @@ def test_ts_slice_to_item():
     sl.to_data_entry()
 
 
+def check_training_validation(
+    train_entry,
+    valid_pair,
+    prediction_length: int,
+    cutoff_date=None,
+    max_history=None,
+    split_offset=None,
+    is_rolling_split=False,
+) -> None:
+    assert train_entry[FieldName.ITEM_ID] == valid_pair[0][FieldName.ITEM_ID]
+    if max_history is None:
+        assert len(train_entry[FieldName.TARGET]) == len(
+            valid_pair[0][FieldName.TARGET]
+        )
+    elif not is_rolling_split:
+        assert (
+            len(valid_pair[0][FieldName.TARGET]) + len(valid_pair[1])
+            == max_history
+        )
+    assert len(valid_pair[1]) == prediction_length
+    train_end = (
+        train_entry[FieldName.START]
+        + len(train_entry[FieldName.TARGET])
+        * train_entry[FieldName.START].freq
+    )
+    if cutoff_date is not None:
+        assert train_end == cutoff_date + train_entry[FieldName.START].freq
+    if split_offset is not None:
+        assert len(train_entry[FieldName.TARGET]) == split_offset
+    assert train_end <= valid_pair[1].index[0]
+    valid_end = (
+        valid_pair[0][FieldName.START]
+        + len(valid_pair[0][FieldName.TARGET])
+        * valid_pair[0][FieldName.START].freq
+    )
+    assert valid_end == valid_pair[1].index[0]
+
+
 def test_splitter():
```

Comment:
This test function got pretty long: you can improve it by encoding a general logic in it, that applies to all possible cases, and have `date, offset, windows, distance, max_history` as arguments to the function. You can then provide all test cases for the above parameters with `pytest`, see for example how [this test](https://github.com/awslabs/gluon-ts/blob/9dd1450bfbf15626ca93ebcb2a73fe74cfe75604/test/test_itertools.py#L60-L80) does it.

I think the testing logic should essentially be:
```python
def test_split_m4_hourly(date, offset, windows, distance, max_history)
    dataset = get_dataset("m4_hourly")
    prediction_length = dataset.metadata.prediction_length
    
    train, test_template = split(dataset.train, date=date, offset=offset)

    train = list(train)
    validation = list(test_template.generate_instances(
        prediction_length=prediction_length,
        windows=windows,
        distance=distance,
        max_history=max_history,
    ))

    assert len(train) == len(dataset)
    assert len(validation) == windows * len(train)

    k = 0
    for train_entry in train:
        for _ in range(windows):
            check_training_validation(...)
            k += 1
```

This way the testing logic is pretty clear and written only once (although some details are hidden inside `check_training_validation`, which is OK).

Also note that I'm using directly `split` in the test body, instead of `DateSplitter` and `OffsetSplitter`, since that makes things simpler.